Implement a Stop the World mechanism for JSC.

MenloDorian opened the pull request:
#### 2e80a204eb24e3950edd84a43ad42fe8aa8ba693
<pre>
Implement a Stop the World mechanism for JSC.
<a href="https://bugs.webkit.org/show_bug.cgi?id=300046">https://bugs.webkit.org/show_bug.cgi?id=300046</a>
<a href="https://rdar.apple.com/161844101">rdar://161844101</a>

Reviewed by Yijia Huang.

This is needed by Global GC, the Wasm Debugger, and the Memory Debugger that we&apos;re exploring
implementing.

The following is documented in VMManager.h, but is replicated here for your convenience in
reading and discovery:

    Understanding Stop the World (STW)
    ==================================

    Intuition on how to think about things?
    =======================================
    The actors in play for a Stop the World story are:
    1. VMManager
    2. VM
    3. a Conductor Agent

    Events / actions involved in the Stop the World story are:
    1. Stop requests (with a given StopReason)
    2. Stop callback handlers

    An intuitive way to think about the Stop the World story is:

    1. VMManager is an abstraction representing the process. There is only 1 singleton
       VMManager instance, and it coordinates the tracking and scheduling of VMs.

    2. VM (and its VMThreadContext) represents a thread. A VM instance may actually
       run on different machine threads at different times (JSC&apos;s API allows this).
       However, from VMManager&apos;s perspective, each VM is like a thread that can be
       suspended / stopped, and resumed.

       FIXME: the current VMManager does NOT yet handle cases where more than one VM
       is run on the same machine thread e.g. one VM1 calls into C++, which in turn
       calls into VM2. VM2 now has control of the CPU, but VMManager does not know
       that VM1 is in a way &quot;deactivated&quot;. This scenario cannot manifest in WebKit
       with web workloads though.

    3. The Conductor Agent is something like a Debugger agent that tells VMManager
       to stop or resume VMs / threads.

    4. Stop requests are like interrupts. A stop being requested is analogous to an
       interrupt firing.

    5. Stop callback handlers are like interrupt handlers that masked out all interrupts
       so that no other interrupts can fire while the current one is being handled.

       When a stop request occurs, VMManager::notifyVMStop() will dispatch a callback
       to the appropriate handler for that request. Similar to the interrupt
       analogy, only one request can be serviced at one time. All other requests
       regardless of priority will be blocked (and held in pending) until the current
       request is done being serviced.

    World Execution Modes
    =====================
    The VMManager has a notion of a world mode (see VMManager::Mode). These modes are:
    1. RunAll - all threads can run or are running.
    2. RunOne - only one thread can run, like when a debugger is single stepping.
    3. Stopping - a stop has been requested, and VMs are in the process of stopping.
    4. Stopped - all threads have been stopped, and the highest priority stop request
                 can now be serviced.

    Querying VMManager Info
    =======================
    VMManager::info() provides a view into a few pieces of VMManager state:
    1. numberOfVMs - the number of VMs that have been constructed and are alive.
    2. numberOfActiveVMs - the number of VMs that are activated i.e. their threads
           have entered the VM. This numberOfActiveVMs is only available while the
           world is NOT in Mode::RunAll (i.e. must be in some form of stoppage).
    3. numberOfStoppedVMs - the number of VMs that have reached the stopping point
           in VMManager::notifyVMStop(). The currently executing targetVM is counted
           as stopped when single stepping in Mode::RunOne.
    3. worldMode - this is the current VM world mode (as described above).

    Currently, this info is mainly used for testing purposes only.

    Initiating Stop the World
    =========================
    Stop the World begins with some agent calling VMManager::requestStopAll() with a
    StopReason. This agent can be from mutator threads or from a helper thread like
    those employed by debuggers.

    More than one agent can request STW at the same time. Hence, there can be multiple
    stop requests queued up while the world is being stopped.

    StopReason and their Priority
    =============================
    Current StopReasons are:
        None - no requests
        GC - requesting stop for Global GC
        WasmDebugger - requesting stop for Wasm Debugger (like Ctrl-C in lldb)
        MemoryDebugger - similar to WasmDebugger, but for the Memory Debugger.

    The priority of these requests are defined by their order of declaration.
    See definition of FOR_EACH_STOP_THE_WORLD_REASON and enum class StopReason.
    StopReason::None is a special case and has no priority.
    StopReason::GC is the current highest priority request.
    StopReason::MemoryDebugger is the current lowest priority request.

    StopReason is synonymous with &quot;StopRequest&quot;.
    From the client&apos;s perspective, it is the reason for a stop request.
    From the VMManager&apos;s perspective, it is the type of stop request.

    Servicing Order: one request at a time
    ======================================
    The order the stop requests came in does not matter. Once the world is finally
    stopped, the higher priority request is serviced first.
    See fetchTopPriorityStopReason() and m_currentStopReason.

    While this request is being serviced, other requests will be ignored. During
    this time of service, new stop requests can be added to m_pendingStopRequestBits,
    but they will be ignored even if they are higher priority. We will service them
    only after the current request has resumed with RunOne or RunAll mode.

    StopTheWorldCallback (i.e. stop request handlers)
    =================================================
    When the world is stopped, VMManager will call back to a request handler
    based on what StopReason is in m_currentStopReason. See VMManager::notifyVMStop().

    The handler for GC should be static but is not currently implemented yet.
    The handlers for WasmDebugger and MemoryDebugger may be overridden. They are
    made to be overrideable only to enable testing. See VMManagerStopTheWorldTest.cpp
    for how they are used in testing.

    Each handler must be of the shape StopTheWorldCallback. See StopTheWorldCallback.h
    The handler will be called with a StopTheWorldEvent. The StopTheWorldEvent indicates
    where the handler is called from. This may have a use later on, but for now, the
    StopTheWorldEvent is only informational.

    After the handler is done, it controls how execution will proceed thereafter by
    returning one of the StopTheWorldCallback return values (see StopTheWorldCallback.h).
    The possible return values are:

    1. STW_CONTINUE()
       - this is only used for testing purposes where we want to loop inside
         VMManager::notifyVMStop() while waiting for more things to handle.
       - VMManager::m_worldMode will remain in Mode::Stopped.
    2. STW_CONTEXT_SWITCH(targetVM)
       - this is used to switch control of the handler to another VM on a different
         thread without resuming any execution. lldb&apos;s  &quot;thread select ...&quot; can be
         implemented this way.
       - VMManager::m_worldMode will remain in Mode::Stopped.
    3. STW_RESUME_ONE()
       - this is used to resume only the current VM thread in RunOne mode. This is
         useful for debuggers that wish to single step in the current VM. It keeps
         other threads paused / stopped while this thread executes. It is up to the
         client to detect potential resource deadlocks (e.g. using a timeout) that
         may arise from only resuming 1 thread.
       - VMManager::m_worldMode will transition from Mode::Stopped to Mode::RunOne.
    4. STW_RESUME_ALL()
       - forces all threads to resume from a stop.
       - VMManager::m_worldMode will transition from Mode::Stopped to Mode::RunAll.
    5. STW_RESUME()
       - Return to whatever run mode we were executing with before the current Stop
         the World request. That may be either Mode::RunOne or Mode::RunAll.
       - This allows the GC to run (with its own Stop the World requests) even while
         while we&apos;re single stepping in a debugger with Mode::RunOne.

    Edge Cases and Special Circumstances
    ====================================
    While in Mode::RunOne, if the VM that is running either exits the VM (aka
    deactivates) or its VM is destructed (aka shutdown), the VMManager will transition
    the world mode back to RunAll (unblocking other VMs and threads)  since the current
    VM is no longer viable for continuing execution.

Tests: Source/JavaScriptCore/API/tests/VMManagerStopTheWorldTest.cpp
       Source/JavaScriptCore/API/tests/VMManagerStopTheWorldTest.h
       Source/JavaScriptCore/API/tests/testapi.c
* Source/JavaScriptCore/API/tests/VMManagerStopTheWorldTest.cpp: Added.
(VMManagerStopTheWorldTest::abortTest):
(VMManagerStopTheWorldTest::threadsList):
(VMManagerStopTheWorldTest::checkpointCallback):
(VMManagerStopTheWorldTest::ensureAliveCallback):
(VMManagerStopTheWorldTest::wasmDebuggerTestCallback):
(VMManagerStopTheWorldTest::memoryDebuggerTestCallback):
(VMManagerStopTheWorldTest::notifyVMDestruction):
(VMManagerStopTheWorldTest::WTF_REQUIRES_LOCK):
(VMManagerStopTheWorldTest::test):
* Source/JavaScriptCore/API/tests/VMManagerStopTheWorldTest.h: Added.
* Source/JavaScriptCore/API/tests/testapi.c:
(main):
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/heap/MarkedBlock.cpp:
(JSC::MarkedBlock::dumpInfoAndCrashForInvalidHandleV2):
* Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
* Source/JavaScriptCore/runtime/JSCConfig.h:
* Source/JavaScriptCore/runtime/StopTheWorldCallback.h: Added.
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::checkStaticAsserts):
(JSC::VM::VM):
(JSC::VM::~VM):
(JSC::VM::hasExceptionsAfterHandlingTraps):
(JSC::VM::throwTerminationException):
(JSC::VM::updateStackLimits):
(JSC::VM::executeEntryScopeServicesOnEntry):
(JSC::VM::executeEntryScopeServicesOnExit):
* Source/JavaScriptCore/runtime/VM.h:
(JSC::VM::hasEntryScopeServiceRequest):
(JSC::VM::offsetOfTraps):
(JSC::VM::fromThreadContext):
(JSC::VM::threadContext):
(JSC::VM::clearLastException):
(JSC::VM::softStackLimit const):
(JSC::VM::addressOfSoftStackLimit):
(JSC::VM::cloopStack):
(JSC::VM::cloopStack const):
(JSC::VM::cloopStackLimit):
(JSC::VM::currentCLoopStackPointer const):
(JSC::VM::traps):
(JSC::VM::traps const):
(JSC::VM::notifyNeedDebuggerBreak):
(JSC::VM::notifyNeedShellTimeoutCheck):
(JSC::VM::notifyNeedTermination):
(JSC::VM::notifyNeedWatchdogCheck):
(JSC::VM::requestStop):
(JSC::VM::cancelStop):
* Source/JavaScriptCore/runtime/VMManager.cpp:
(JSC::VMManager::singleton):
(JSC::VMThreadContext::VMThreadContext):
(JSC::VMThreadContext::~VMThreadContext):
(JSC::WTF_REQUIRES_LOCK):
(JSC::VMManager::findMatchingVMImpl):
(JSC::VMManager::forEachVMImpl):
(JSC::VMManager::forEachVMWithTimeoutImpl):
(JSC::VMManager::info):
(JSC::VMManager::setWasmDebuggerCallback):
(JSC::VMManager::setMemoryDebuggerCallback):
(JSC::VMManager::requestStopAllInternal):
(JSC::VMManager::requestResumeAllInternal):
(JSC::VMManager::notifyVMStop):
(JSC::VMManager::notifyVMConstruction):
(JSC::VMManager::notifyVMDestruction):
(JSC::VMManager::notifyVMActivation):
(JSC::VMManager::notifyVMDeactivation):
(JSC::VMManager::handleVMDestructionWhileWorldStopped):
(JSC::VMManager::add): Deleted.
(JSC::VMManager::remove): Deleted.
* Source/JavaScriptCore/runtime/VMManager.h:
(JSC::VMManager::numberOfVMs):
(JSC::VMManager::requestStopAll):
(JSC::VMManager::requestResumeAll):
(JSC::VMManager::findMatchingVM):
(JSC::VMManager::forEachVM):
(JSC::VMManager::forEachVMWithTimeout):
(JSC::VMManager::hasPendingStopRequests const):
* Source/JavaScriptCore/runtime/VMThreadContext.h: Added.
(JSC::VMThreadContext::next const):
(JSC::VMThreadContext::traps):
(JSC::VMThreadContext::traps const):
(JSC::VMThreadContext::offsetOfTraps):
* Source/JavaScriptCore/runtime/VMTraps.cpp:
(JSC::VMTraps::handleTraps):
* Source/JavaScriptCore/runtime/VMTraps.h:
* Source/JavaScriptCore/runtime/VMTrapsInlines.h:
(JSC::VMTraps::vm const):
* Source/JavaScriptCore/shell/CMakeLists.txt:
* Source/JavaScriptCore/wasm/WasmCallee.cpp:
(JSC::Wasm::Callee::reportToVMsForDestruction):

Canonical link: <a href="https://commits.webkit.org/301027@main">https://commits.webkit.org/301027@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e27d2fc058b988ceefaae5f6df20b2b2bfcfe7b5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124672 "9 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44344 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35078 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131513 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/76599 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/9c332daf-97bf-4c9f-a5e3-d140404243f9) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45048 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52914 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94853 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62912 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/f38e3b28-9e05-4368-8535-6fe7bb537f09) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127626 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35924 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111496 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75423 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/6512cdd7-347c-40ae-84cd-acb60e147b7e) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34854 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29651 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74993 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/116777 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105681 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29882 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134179 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/123192 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51520 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39336 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103327 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51932 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107717 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103101 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48476 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26739 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/48476 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19559 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51394 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/57191 "Built successfully") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/156296 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50787 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/156296 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54143 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52482 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

hyjorc1 commented on Source/JavaScriptCore/runtime/VMManager.h:
```diff
@@ -36,50 +40,341 @@ namespace JSC {
+// 3. a Conductor Agent
+//
+// Events / actions involved in the Stop the World story are:
+// 1. StopRequests
```
Maybe `RequestAgent` is a better name? Since it reads naturally for
```
VMManager::requestStopAll(RequestAgent::GC); // request stop all from agent GC
VMManager::requestResumeAll(RequestAgent::GC); // request resume all from agent GC
```

MenloDorian replied:
I debated between calling it `StopReason` vs `StopRequest`.  From the client side, "reason" reads better, but internally, this reason gets put into a `m_pendingRequestBits` ... which implies the value is a "request", and the implementation reads better when the values it operates on of a consistent type.  That said, I think I can smooth this away with a comment at the point where I "cast" a reason into a request.  I'll go ahead and make that change.

I prefer not to call it an "agent" as that seems even further away from the concept of a "request", which is what we operate on in the internal implementation.

hyjorc1 commented on Source/JavaScriptCore/runtime/VMManager.cpp:
```diff
@@ -107,18 +106,426 @@ VM* VMManager::findMatchingVMImpl(const ScopedLambda<VMManager::TestCallback>& t
+            }
+        }
+
+        if (status.second && status.second != STW_RESUME_ALL_TOKEN && status.second != m_targetVM) {
```
For the switch context case, it's switch the target VM which serves the stop request right? Then, does it work correctly with `STW_CONTEXT_SWITCH(same_vm)` and `STW_RESUME_ONE(diff_vm)` w.r.t `m_targetVM`?

MenloDorian replied:
Yes, you can context switch to the same VM or a different VM.  You can also resume RunOne on the same VM or a different VM.  The included test may or may not context switch to the same VM depending on which VM it stopped in last.  However, in my debugging, I've seen it context switch to the same VM and just works.  The test also does not exercise resume  RunOne on the same VM.  However, the mechanism of VM selection is identical with the context switch one.  So, it should just work.

hyjorc1 commented on Source/JavaScriptCore/runtime/VMManager.cpp:
```diff
@@ -107,18 +106,426 @@ VM* VMManager::findMatchingVMImpl(const ScopedLambda<VMManager::TestCallback>& t
+            if (status.second) {
+                if (status.second == STW_RESUME_ALL_TOKEN)
+                    m_useRunOneMode = false; // Clearing to ensure we STW_RESUME_ALL.
+                else if (status.second)
```
Should be `else`?

hyjorc1 replied:
Or should be `m_useRunOneMode = status.second != STW_RESUME_ALL_TOKEN;`.

MenloDorian replied:
You are correct.  This if + if else was residual from old logic before I added the outer `if (status.second)`.  It can now be simplified.  I'll make the change.

hyjorc1 commented on Source/JavaScriptCore/runtime/VMManager.h:
```diff
@@ -36,50 +40,346 @@ namespace JSC {
+//    VMManager instance, and it coordinates the tracking and scheduling of VMs.
+//
+// 2. VM (and its VMThreadContext) represents a thread. A VM instance may actually
+//    run on different machine threads at differen times (JSC's API allows this).
```
`differen` -> `different`

hyjorc1 commented on Source/JavaScriptCore/API/tests/VMManagerStopTheWorldTest.cpp:
```diff
@@ -0,0 +1,1400 @@
+ customize the callbacks for these debuggers.
+ - see wasmDebuggerTestCallback() and memoryDebuggerTestCallback().
+
+ In this test, the test WasmDebugger will execise STW feature like the real
```
`execise` -> `exercise`

hyjorc1 commented on Source/JavaScriptCore/runtime/VM.cpp:
```diff
@@ -212,6 +212,25 @@ void VM::computeCanUseJIT()
+{
+    // VM registration is done in the instantiation of its VMThreadContext.
+    //
+    // VM registration with the VMManager can only be done after m_apiLock is initiaized
```
`initiaized` -> `initialized`

hyjorc1 commented on Source/JavaScriptCore/runtime/VMThreadContext.h:
```diff
@@ -0,0 +1,61 @@
+    enum class State : uint8_t {
+        Run,
+        Stop,
+    };
```
This enum class seems not be used in this patch, right?

hyjorc1 commented on Source/JavaScriptCore/runtime/VMManager.cpp:
```diff
@@ -107,18 +106,426 @@ VM* VMManager::findMatchingVMImpl(const ScopedLambda<VMManager::TestCallback>& t
+void VMManager::decrementActiveVMs(VM& vm) WTF_REQUIRES_LOCK(m_worldLock)
+{
+    // We only need to track m_numberOfActiveVMs changes if we're in RunOne
+    // mode. If we're running be cause the world was resumed with RunAll,
```
`be cause` -> `because`

hyjorc1 commented on Source/JavaScriptCore/runtime/VMManager.cpp:
```diff
@@ -107,18 +106,426 @@ VM* VMManager::findMatchingVMImpl(const ScopedLambda<VMManager::TestCallback>& t
+            // RunOne mode allows execution of 1 VM without resumeTheWorld(). We did not clear
+            // the m_hasBeenCountedAsActive flags on each VM on resuming with RunOne. As a
+            // result, m_numberOfActiveVMs is still valid in RunOne mode. We don't want
+            // want to reset m_numberOfActiveVMs to 0 here because we won't be re-calculating
```
`want want` -> `want`

hyjorc1 commented on Source/JavaScriptCore/runtime/VMManager.cpp:
```diff
@@ -107,18 +106,426 @@ VM* VMManager::findMatchingVMImpl(const ScopedLambda<VMManager::TestCallback>& t
+                if (m_currentStopReason == StopReason::None)
+                    return false;
+
+                // 3. We have a STW request. Stop is not all active VMs are at the stopping point yet,
```
`Stop is not` -> `Stop if not`

MenloDorian replied:
Fixed.  Should be "If not".

hyjorc1 commented on Source/JavaScriptCore/runtime/VMManager.h:
```diff
@@ -36,50 +40,346 @@ namespace JSC {
+
+    // Indicates the targetVM that will service the StopTheWorld request, or the targetVM that may
+    // continue running in RunOne mode.
+    // Can obly be written to while holding the m_worldLock.
```
`obly` -> `only`

hyjorc1 reviewed (approved):
Nice! r=me if EWS is happy.

Commit:
Implement a Stop the World mechanism for JSC.
https://bugs.webkit.org/show_bug.cgi?id=300046
rdar://161844101

Reviewed by Yijia Huang.

This is needed by Global GC, the Wasm Debugger, and the Memory Debugger that we're exploring
implementing.

The following is documented in VMManager.h, but is replicated here for your convenience in
reading and discovery:

    Understanding Stop the World (STW)
    ==================================

    Intuition on how to think about things?
    =======================================
    The actors in play for a Stop the World story are:
    1. VMManager
    2. VM
    3. a Conductor Agent

    Events / actions involved in the Stop the World story are:
    1. Stop requests (with a given StopReason)
    2. Stop callback handlers

    An intuitive way to think about the Stop the World story is:

    1. VMManager is an abstraction representing the process. There is only 1 singleton
       VMManager instance, and it coordinates the tracking and scheduling of VMs.

    2. VM (and its VMThreadContext) represents a thread. A VM instance may actually
       run on different machine threads at different times (JSC's API allows this).
       However, from VMManager's perspective, each VM is like a thread that can be
       suspended / stopped, and resumed.

       FIXME: the current VMManager does NOT yet handle cases where more than one VM
       is run on the same machine thread e.g. one VM1 calls into C++, which in turn
       calls into VM2. VM2 now has control of the CPU, but VMManager does not know
       that VM1 is in a way "deactivated". This scenario cannot manifest in WebKit
       with web workloads though.

    3. The Conductor Agent is something like a Debugger agent that tells VMManager
       to stop or resume VMs / threads.

    4. Stop requests are like interrupts. A stop being requested is analogous to an
       interrupt firing.

    5. Stop callback handlers are like interrupt handlers that masked out all interrupts
       so that no other interrupts can fire while the current one is being handled.

       When a stop request occurs, VMManager::notifyVMStop() will dispatch a callback
       to the appropriate handler for that request. Similar to the interrupt
       analogy, only one request can be serviced at one time. All other requests
       regardless of priority will be blocked (and held in pending) until the current
       request is done being serviced.

    World Execution Modes
    =====================
    The VMManager has a notion of a world mode (see VMManager::Mode). These modes are:
    1. RunAll - all threads can run or are running.
    2. RunOne - only one thread can run, like when a debugger is single stepping.
    3. Stopping - a stop has been requested, and VMs are in the process of stopping.
    4. Stopped - all threads have been stopped, and the highest priority stop request
                 can now be serviced.

    Querying VMManager Info
    =======================
    VMManager::info() provides a view into a few pieces of VMManager state:
    1. numberOfVMs - the number of VMs that have been constructed and are alive.
    2. numberOfActiveVMs - the number of VMs that are activated i.e. their threads
           have entered the VM. This numberOfActiveVMs is only available while the
           world is NOT in Mode::RunAll (i.e. must be in some form of stoppage).
    3. numberOfStoppedVMs - the number of VMs that have reached the stopping point
           in VMManager::notifyVMStop(). The currently executing targetVM is counted
           as stopped when single stepping in Mode::RunOne.
    3. worldMode - this is the current VM world mode (as described above).

    Currently, this info is mainly used for testing purposes only.

    Initiating Stop the World
    =========================
    Stop the World begins with some agent calling VMManager::requestStopAll() with a
    StopReason. This agent can be from mutator threads or from a helper thread like
    those employed by debuggers.

    More than one agent can request STW at the same time. Hence, there can be multiple
    stop requests queued up while the world is being stopped.

    StopReason and their Priority
    =============================
    Current StopReasons are:
        None - no requests
        GC - requesting stop for Global GC
        WasmDebugger - requesting stop for Wasm Debugger (like Ctrl-C in lldb)
        MemoryDebugger - similar to WasmDebugger, but for the Memory Debugger.

    The priority of these requests are defined by their order of declaration.
    See definition of FOR_EACH_STOP_THE_WORLD_REASON and enum class StopReason.
    StopReason::None is a special case and has no priority.
    StopReason::GC is the current highest priority request.
    StopReason::MemoryDebugger is the current lowest priority request.

    StopReason is synonymous with "StopRequest".
    From the client's perspective, it is the reason for a stop request.
    From the VMManager's perspective, it is the type of stop request.

    Servicing Order: one request at a time
    ======================================
    The order the stop requests came in does not matter. Once the world is finally
    stopped, the higher priority request is serviced first.
    See fetchTopPriorityStopReason() and m_currentStopReason.

    While this request is being serviced, other requests will be ignored. During
    this time of service, new stop requests can be added to m_pendingStopRequestBits,
    but they will be ignored even if they are higher priority. We will service them
    only after the current request has resumed with RunOne or RunAll mode.

    StopTheWorldCallback (i.e. stop request handlers)
    =================================================
    When the world is stopped, VMManager will call back to a request handler
    based on what StopReason is in m_currentStopReason. See VMManager::notifyVMStop().

    The handler for GC should be static but is not currently implemented yet.
    The handlers for WasmDebugger and MemoryDebugger may be overridden. They are
    made to be overrideable only to enable testing. See VMManagerStopTheWorldTest.cpp
    for how they are used in testing.

    Each handler must be of the shape StopTheWorldCallback. See StopTheWorldCallback.h
    The handler will be called with a StopTheWorldEvent. The StopTheWorldEvent indicates
    where the handler is called from. This may have a use later on, but for now, the
    StopTheWorldEvent is only informational.

    After the handler is done, it controls how execution will proceed thereafter by
    returning one of the StopTheWorldCallback return values (see StopTheWorldCallback.h).
    The possible return values are:

    1. STW_CONTINUE()
       - this is only used for testing purposes where we want to loop inside
         VMManager::notifyVMStop() while waiting for more things to handle.
       - VMManager::m_worldMode will remain in Mode::Stopped.
    2. STW_CONTEXT_SWITCH(targetVM)
       - this is used to switch control of the handler to another VM on a different
         thread without resuming any execution. lldb's  "thread select ..." can be
         implemented this way.
       - VMManager::m_worldMode will remain in Mode::Stopped.
    3. STW_RESUME_ONE()
       - this is used to resume only the current VM thread in RunOne mode. This is
         useful for debuggers that wish to single step in the current VM. It keeps
         other threads paused / stopped while this thread executes. It is up to the
         client to detect potential resource deadlocks (e.g. using a timeout) that
         may arise from only resuming 1 thread.
       - VMManager::m_worldMode will transition from Mode::Stopped to Mode::RunOne.
    4. STW_RESUME_ALL()
       - forces all threads to resume from a stop.
       - VMManager::m_worldMode will transition from Mode::Stopped to Mode::RunAll.
    5. STW_RESUME()
       - Return to whatever run mode we were executing with before the current Stop
         the World request. That may be either Mode::RunOne or Mode::RunAll.
       - This allows the GC to run (with its own Stop the World requests) even while
         while we're single stepping in a debugger with Mode::RunOne.

    Edge Cases and Special Circumstances
    ====================================
    While in Mode::RunOne, if the VM that is running either exits the VM (aka
    deactivates) or its VM is destructed (aka shutdown), the VMManager will transition
    the world mode back to RunAll (unblocking other VMs and threads)  since the current
    VM is no longer viable for continuing execution.

Tests: Source/JavaScriptCore/API/tests/VMManagerStopTheWorldTest.cpp
       Source/JavaScriptCore/API/tests/VMManagerStopTheWorldTest.h
       Source/JavaScriptCore/API/tests/testapi.c
* Source/JavaScriptCore/API/tests/VMManagerStopTheWorldTest.cpp: Added.
(VMManagerStopTheWorldTest::abortTest):
(VMManagerStopTheWorldTest::threadsList):
(VMManagerStopTheWorldTest::checkpointCallback):
(VMManagerStopTheWorldTest::ensureAliveCallback):
(VMManagerStopTheWorldTest::wasmDebuggerTestCallback):
(VMManagerStopTheWorldTest::memoryDebuggerTestCallback):
(VMManagerStopTheWorldTest::notifyVMDestruction):
(VMManagerStopTheWorldTest::WTF_REQUIRES_LOCK):
(VMManagerStopTheWorldTest::test):
* Source/JavaScriptCore/API/tests/VMManagerStopTheWorldTest.h: Added.
* Source/JavaScriptCore/API/tests/testapi.c:
(main):
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/heap/MarkedBlock.cpp:
(JSC::MarkedBlock::dumpInfoAndCrashForInvalidHandleV2):
* Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
* Source/JavaScriptCore/runtime/JSCConfig.h:
* Source/JavaScriptCore/runtime/StopTheWorldCallback.h: Added.
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::checkStaticAsserts):
(JSC::VM::VM):
(JSC::VM::~VM):
(JSC::VM::hasExceptionsAfterHandlingTraps):
(JSC::VM::throwTerminationException):
(JSC::VM::updateStackLimits):
(JSC::VM::executeEntryScopeServicesOnEntry):
(JSC::VM::executeEntryScopeServicesOnExit):
* Source/JavaScriptCore/runtime/VM.h:
(JSC::VM::hasEntryScopeServiceRequest):
(JSC::VM::offsetOfTraps):
(JSC::VM::fromThreadContext):
(JSC::VM::threadContext):
(JSC::VM::clearLastException):
(JSC::VM::softStackLimit const):
(JSC::VM::addressOfSoftStackLimit):
(JSC::VM::cloopStack):
(JSC::VM::cloopStack const):
(JSC::VM::cloopStackLimit):
(JSC::VM::currentCLoopStackPointer const):
(JSC::VM::traps):
(JSC::VM::traps const):
(JSC::VM::notifyNeedDebuggerBreak):
(JSC::VM::notifyNeedShellTimeoutCheck):
(JSC::VM::notifyNeedTermination):
(JSC::VM::notifyNeedWatchdogCheck):
(JSC::VM::requestStop):
(JSC::VM::cancelStop):
* Source/JavaScriptCore/runtime/VMManager.cpp:
(JSC::VMManager::singleton):
(JSC::VMThreadContext::VMThreadContext):
(JSC::VMThreadContext::~VMThreadContext):
(JSC::WTF_REQUIRES_LOCK):
(JSC::VMManager::findMatchingVMImpl):
(JSC::VMManager::forEachVMImpl):
(JSC::VMManager::forEachVMWithTimeoutImpl):
(JSC::VMManager::info):
(JSC::VMManager::setWasmDebuggerCallback):
(JSC::VMManager::setMemoryDebuggerCallback):
(JSC::VMManager::requestStopAllInternal):
(JSC::VMManager::requestResumeAllInternal):
(JSC::VMManager::notifyVMStop):
(JSC::VMManager::notifyVMConstruction):
(JSC::VMManager::notifyVMDestruction):
(JSC::VMManager::notifyVMActivation):
(JSC::VMManager::notifyVMDeactivation):
(JSC::VMManager::handleVMDestructionWhileWorldStopped):
(JSC::VMManager::add): Deleted.
(JSC::VMManager::remove): Deleted.
* Source/JavaScriptCore/runtime/VMManager.h:
(JSC::VMManager::numberOfVMs):
(JSC::VMManager::requestStopAll):
(JSC::VMManager::requestResumeAll):
(JSC::VMManager::findMatchingVM):
(JSC::VMManager::forEachVM):
(JSC::VMManager::forEachVMWithTimeout):
(JSC::VMManager::hasPendingStopRequests const):
* Source/JavaScriptCore/runtime/VMThreadContext.h: Added.
(JSC::VMThreadContext::next const):
(JSC::VMThreadContext::traps):
(JSC::VMThreadContext::traps const):
(JSC::VMThreadContext::offsetOfTraps):
* Source/JavaScriptCore/runtime/VMTraps.cpp:
(JSC::VMTraps::handleTraps):
* Source/JavaScriptCore/runtime/VMTraps.h:
* Source/JavaScriptCore/runtime/VMTrapsInlines.h:
(JSC::VMTraps::vm const):
* Source/JavaScriptCore/shell/CMakeLists.txt:
* Source/JavaScriptCore/wasm/WasmCallee.cpp:
(JSC::Wasm::Callee::reportToVMsForDestruction):

Canonical link: https://commits.webkit.org/301027@main
